rfcs/convergence/react-focus-management.md

```diff
@@ -0,0 +1,169 @@
+
+#### Hooks
+
+This spec proposes an API based on React hooks pattern. [ability-helpers](https://github.com/microsoft/ability-helpers) works mainly on adding `data-*` attributes to DOM elements to apply configurable behaviours to the 'groups' of focusables.
+
+However some internal behaviours such as `modalizer` or `deloser` require DOM elements during the configuration phase, which will be most easily configured in React using `refs`.
+
+Therefore it must be noted that **although some hooks might not require ref interaction and can simply return attributes to be spread, in an effort to maintain consistency refs are used, which fits the Fluent component design pattern which enforces ref forwarding**
+
+```typescript
+// Basic up/down keyboard navigation
+const ref = useKeyboardNavigationGroup();
+
+<div ref={ref}>{props.children}</div>;
+```
+
+```typescript
+const ref = React.useReact<HTMLElement>();
+const {
+  findFirst,
+  findLast,
+  findAll,
+} = useFocusable(ref); // All operations are with respect to the given DOM container
+
+findFirst() // first focusable
+findLast() // first last focusable
+findAll(condition: () => true) // first all focusables with condition
+
+<div ref={ref}>{props.children}</div>
+```
+
+#### useFocusable
+
+A hook that exposes helper methods to traverse focusable elements in the context of a DOM element.
+
+The below table lists the helper functions to support:
+
+| Name      | Value         | Options                  | Description                                         |
+| --------- | ------------- | ------------------------ | --------------------------------------------------- |
+| findFirst | HTMLElement   |                          | Finds the first focusable element                   |
+| findLast  | HTMLElement   |                          | Finds the last focusable element                    |
+| findAll   | HTMLElement[] | boolean matcher function | Finds all focusable elements that match a condition |
+
+#### useKeyboardNagiationGroup
+
+A hook that returns a `ref` to be assigned to the container of a group of focusable elements that should support keyboard navigation
+
+The below table lists the configuration options that the hook should support:
+
+| Name          | Value                  | Description                                                                                               |
+| ------------- | ---------------------- | --------------------------------------------------------------------------------------------------------- |
+| axis          | vertical \| horizontal | Indicates which direction keyboard navigation should use                                                  |
+| enableHomeEnd | boolean                | Enables the use of `Home` and `End` keys to navigate to first and last elements of the group respectively |
+| cyclic        | boolean                | Indicates that navigation should not stop at first and last elements respectively                         |
```

Comment:
FYI: `FocusZone` uses `isCircularNavigation`, so may be `circular` is also an option there...

Reply:
updated